In [1]:
# 2-10 여러 데이터셋 뜯어보기

In [34]:
import pandas as pd
from sklearn.model_selection import train_test_split
# 오차행렬 불러오기
from sklearn.metrics import confusion_matrix
from sklearn.metrics import precision_score
from sklearn.metrics import f1_score
from sklearn.metrics import recall_score

### 2-11. 프로젝트(1) load_digits: 손글씨 분류하기
---
- Feature Data 지정하기
- Label Data 지정하기
- Target Names 출력해 보기
- 데이터 Describe 해 보기
---
1. module import
2. data prepare
3. understainding of data
4. train, test split
5. train data with various models
---
- decision tree
- random forest
- svm
- sgd classifier
- logistic regression
6. evaluate model

In [3]:
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [4]:
# decision tree
from sklearn.tree import DecisionTreeClassifier

# data prepare
digits = load_digits()
digits_data = digits.data
digits_label = digits.target

# data isolation
x_train, x_test, y_train, y_test = train_test_split(digits_data,
                                                   digits_label,
                                                   test_size=0.2,
                                                   random_state=32)

# model train & predict
dst_model = DecisionTreeClassifier(random_state=32)
dst_model.fit(x_train, y_train)
y_pred = dst_model.predict(x_test)

accuracy = accuracy_score(y_test, y_pred)

print(classification_report(y_test, y_pred))
print(accuracy)


              precision    recall  f1-score   support

           0       1.00      0.89      0.94        38
           1       0.75      0.83      0.79        36
           2       0.75      0.84      0.79        32
           3       0.92      0.86      0.89        56
           4       0.85      0.90      0.88        31
           5       0.92      0.97      0.95        36
           6       0.97      0.94      0.96        34
           7       0.94      0.88      0.91        34
           8       0.88      0.78      0.82        27
           9       0.79      0.83      0.81        36

    accuracy                           0.88       360
   macro avg       0.88      0.87      0.87       360
weighted avg       0.88      0.88      0.88       360

0.875


In [5]:
# randomforest
from sklearn.ensemble import RandomForestClassifier

# digits = load_digits()
# digits_data = digits.data
# digits_label = digits.target

x_train, x_test, y_train, y_test = train_test_split(digits_data,
                                                   digits_label,
                                                   test_size=0.2,
                                                   random_state=32)

rf = RandomForestClassifier(random_state=32)
rf.fit(x_train, y_train)
y_pred = rf.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        38
           1       0.97      1.00      0.99        36
           2       0.97      1.00      0.98        32
           3       1.00      1.00      1.00        56
           4       1.00      0.97      0.98        31
           5       1.00      0.97      0.99        36
           6       1.00      1.00      1.00        34
           7       0.97      1.00      0.99        34
           8       0.96      0.89      0.92        27
           9       0.97      1.00      0.99        36

    accuracy                           0.99       360
   macro avg       0.98      0.98      0.98       360
weighted avg       0.99      0.99      0.99       360

0.9861111111111112


In [41]:
# svm
from sklearn import svm
# sgd
# logistic

# digits = load_digits()
# digits_data = digits.data
# digits_label = digits.target

x_train, x_test, y_train, y_test = train_test_split(digits_data,
                                                   digits_label,
                                                   test_size=0.2,
                                                   random_state=32)

svm = svm.SVC()
svm.fit(x_train, y_train)
y_pred = svm.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00        38
           1       1.00      1.00      1.00        36
           2       1.00      1.00      1.00        32
           3       1.00      1.00      1.00        56
           4       1.00      0.97      0.98        31
           5       1.00      0.97      0.99        36
           6       1.00      1.00      1.00        34
           7       1.00      1.00      1.00        34
           8       1.00      1.00      1.00        27
           9       0.95      1.00      0.97        36

    accuracy                           0.99       360
   macro avg       0.99      0.99      0.99       360
weighted avg       0.99      0.99      0.99       360

accuracy: 0.9944444444444445


ValueError: Target is multiclass but average='binary'. Please choose another average setting, one of [None, 'micro', 'macro', 'weighted'].

In [7]:
# sgd
from sklearn.linear_model import SGDClassifier
# logistic

# digits = load_digits()
# digits_data = digits.data
# digits_label = digits.target

x_train, x_test, y_train, y_test = train_test_split(digits_data,
                                                   digits_label,
                                                   test_size=0.2,
                                                   random_state=32)

sgd = SGDClassifier()
sgd.fit(x_train, y_train)
y_pred = sgd.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.97      0.99        38
           1       0.97      0.92      0.94        36
           2       1.00      1.00      1.00        32
           3       0.98      0.91      0.94        56
           4       1.00      0.97      0.98        31
           5       0.88      0.97      0.92        36
           6       0.97      1.00      0.99        34
           7       1.00      1.00      1.00        34
           8       0.93      0.96      0.95        27
           9       0.92      0.97      0.95        36

    accuracy                           0.96       360
   macro avg       0.96      0.97      0.97       360
weighted avg       0.97      0.96      0.96       360

0.9638888888888889


In [8]:
# logistic
from sklearn.linear_model import LogisticRegression

digits = load_digits()
digits_data = digits.data
digits_label = digits.target

x_train, x_test, y_train, y_test = train_test_split(digits_data,
                                                   digits_label,
                                                   test_size=0.2,
                                                   random_state=32)

lr = LogisticRegression() # max_iter=500
lr.fit(x_train, y_train)
y_pred = lr.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))


              precision    recall  f1-score   support

           0       1.00      0.95      0.97        38
           1       0.95      0.97      0.96        36
           2       1.00      0.97      0.98        32
           3       1.00      0.98      0.99        56
           4       1.00      0.97      0.98        31
           5       0.94      0.92      0.93        36
           6       1.00      1.00      1.00        34
           7       1.00      0.97      0.99        34
           8       0.96      1.00      0.98        27
           9       0.88      1.00      0.94        36

    accuracy                           0.97       360
   macro avg       0.97      0.97      0.97       360
weighted avg       0.97      0.97      0.97       360

0.9722222222222222


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### 

### 2-11 프로젝트(2) - wine data 분류하기

In [9]:
from sklearn.datasets import load_wine

from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [10]:
wine = load_wine()
wine_data = wine.data
wine_label = wine.target

In [11]:
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size=0.2,
                                                   random_state=32)

In [12]:
# decision tree
from sklearn.tree import DecisionTreeClassifier

dst_model = DecisionTreeClassifier(random_state=32)
dst_model.fit(x_train, y_train)
y_pred = dst_model.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      1.00      0.97        16
           1       0.89      0.80      0.84        10
           2       0.90      0.90      0.90        10

    accuracy                           0.92        36
   macro avg       0.91      0.90      0.90        36
weighted avg       0.92      0.92      0.91        36

0.9166666666666666


In [13]:
# randomforest
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size=0.2,
                                                   random_state=32)

from sklearn.ensemble import RandomForestClassifier
rf_model = RandomForestClassifier(random_state=32)
rf_model.fit(x_train, y_train)
y_pred=rf_model.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        16
           1       1.00      0.90      0.95        10
           2       0.91      1.00      0.95        10

    accuracy                           0.97        36
   macro avg       0.97      0.97      0.97        36
weighted avg       0.97      0.97      0.97        36

0.9722222222222222


In [14]:
# svm
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size=0.2,
                                                   random_state=32)

from sklearn import svm
svm_model = svm.SVC()
svm_model.fit(x_train,y_train)
y_pred = svm_model.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.93      0.81      0.87        16
           1       0.41      0.90      0.56        10
           2       0.00      0.00      0.00        10

    accuracy                           0.61        36
   macro avg       0.45      0.57      0.48        36
weighted avg       0.53      0.61      0.54        36

0.6111111111111112


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [15]:
#sgd
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size=0.2,
                                                   random_state=32)

from sklearn.linear_model import SGDClassifier
sgd_model = SGDClassifier()
sgd_model.fit(x_train, y_train)
y_pred=sgd_model.predict(x_test)

print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      0.69      0.81        16
           1       0.40      1.00      0.57        10
           2       0.00      0.00      0.00        10

    accuracy                           0.58        36
   macro avg       0.47      0.56      0.46        36
weighted avg       0.56      0.58      0.52        36

0.5833333333333334


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/metrics/_classification.py:1221: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [16]:
#logistic_regression
x_train, x_test, y_train, y_test = train_test_split(wine_data,
                                                   wine_label,
                                                   test_size=0.2,
                                                   random_state=32)

from sklearn.linear_model import LogisticRegression
lgr_model = LogisticRegression()
lgr_model.fit(x_train, y_train)
y_pred = lgr_model.predict(x_test)


print(classification_report(y_test, y_pred))
print(accuracy_score(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.94      1.00      0.97        16
           1       0.89      0.80      0.84        10
           2       0.90      0.90      0.90        10

    accuracy                           0.92        36
   macro avg       0.91      0.90      0.90        36
weighted avg       0.92      0.92      0.91        36

0.9166666666666666


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### 2-12. 프로젝트(3) - breast cancer 

In [17]:
from sklearn.datasets import load_breast_cancer

from sklearn.model_selection import train_test_split
from sklearn.metrics import classification_report
from sklearn.metrics import accuracy_score

In [18]:
# data 준비
bcancer = load_breast_cancer()
bcancer_data = bcancer.data
bcancer_label = bcancer.target

In [28]:
# decision tree
from sklearn.tree import DecisionTreeClassifier

x_train, x_test, y_train, y_test = train_test_split(bcancer_data, 
                                                    bcancer_label, 
                                                    test_size=0.2, 
                                                    random_state=32)

dst_model = DecisionTreeClassifier(random_state=32)
dst_model.fit(x_train,y_train)
y_pred=dst_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("accuracy:",accuracy_score(y_test,y_pred))
print("precision_score",precision_score(y_test,y_pred))
print("f1 score:", f1_score(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.84      0.84      0.84        44
           1       0.90      0.90      0.90        70

    accuracy                           0.88       114
   macro avg       0.87      0.87      0.87       114
weighted avg       0.88      0.88      0.88       114

accuracy: 0.8771929824561403
precision_score 0.9
f1 score: 0.9


In [40]:
# randomforest
from sklearn.ensemble import RandomForestClassifier

x_train, x_test, y_train, y_test=train_test_split(bcancer_data,
                                                 bcancer_label,
                                                 test_size=0.2,
                                                 random_state=32)
rf_model=RandomForestClassifier(random_state=32)
rf_model.fit(x_train,y_train)
y_pred=rf_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("accuracy:",accuracy_score(y_test,y_pred))
print("precision_score",precision_score(y_test,y_pred))
print("recall:", recall_score(y_test,y_pred))
print("f1 score:", f1_score(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.91      0.93      0.92        44
           1       0.96      0.94      0.95        70

    accuracy                           0.94       114
   macro avg       0.93      0.94      0.94       114
weighted avg       0.94      0.94      0.94       114

accuracy: 0.9385964912280702
precision_score 0.9565217391304348
recall: 0.9428571428571428
f1 score: 0.9496402877697843


In [39]:
# svm
from sklearn import svm

x_train, x_test, y_train, y_test=train_test_split(bcancer_data,
                                                 bcancer_label,
                                                 test_size=0.2,
                                                 random_state=32)
svm_model=svm.SVC()
svm_model.fit(x_train,y_train)
y_pred=svm_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("accuracy:",accuracy_score(y_test,y_pred))
print("precision_score",precision_score(y_test,y_pred))
print("recall:", recall_score(y_test,y_pred))
print("f1 score:", f1_score(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.91      0.73      0.81        44
           1       0.85      0.96      0.90        70

    accuracy                           0.87       114
   macro avg       0.88      0.84      0.85       114
weighted avg       0.87      0.87      0.86       114

accuracy: 0.868421052631579
precision_score 0.8481012658227848
recall: 0.9571428571428572
f1 score: 0.8993288590604026


In [36]:
# sgd
from sklearn.linear_model import SGDClassifier

x_train, x_test, y_train, y_test=train_test_split(bcancer_data,
                                                 bcancer_label,
                                                 test_size=0.2,
                                                 random_state=32)
sgd_model=SGDClassifier()
sgd_model.fit(x_train,y_train)
y_pred=sgd_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("accuracy:",accuracy_score(y_test,y_pred))
print("precision_score",precision_score(y_test,y_pred))
print("recall:", recall_score(y_test,y_pred))
print("f1 score:", f1_score(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.63      0.98      0.77        44
           1       0.98      0.64      0.78        70

    accuracy                           0.77       114
   macro avg       0.81      0.81      0.77       114
weighted avg       0.84      0.77      0.77       114

accuracy: 0.7719298245614035
precision_score 0.9782608695652174
recall: 0.6428571428571429
f1 score: 0.7758620689655172


In [38]:
# logistic regression
from sklearn.linear_model import LogisticRegression

x_train, x_test, y_train, y_test=train_test_split(bcancer_data,
                                                 bcancer_label,
                                                 test_size=0.2,
                                                 random_state=32)
lgr_model=LogisticRegression()
lgr_model.fit(x_train,y_train)
y_pred=lgr_model.predict(x_test)

print(classification_report(y_test, y_pred))
print("accuracy:",accuracy_score(y_test,y_pred))
print("precision_score",precision_score(y_test,y_pred))
print("recall:", recall_score(y_test,y_pred))
print("f1 score:", f1_score(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.88      0.86      0.87        44
           1       0.92      0.93      0.92        70

    accuracy                           0.90       114
   macro avg       0.90      0.90      0.90       114
weighted avg       0.90      0.90      0.90       114

accuracy: 0.9035087719298246
precision_score 0.9154929577464789
recall: 0.9285714285714286
f1 score: 0.921985815602837


/home/aiffel/anaconda3/envs/aiffel/lib/python3.7/site-packages/sklearn/linear_model/_logistic.py:764: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


### cancer data analysis

1. accuracy per model
dst - 0.875
rf - 0.986
svm = 0.99
sgd = 0.96
logistic = 0.97

2. precision per model


3. f1score per model